frontend/stmt_inferrer.py

```diff
@@ -0,0 +1,301 @@
+
+    Limitation:
+        - The iterable can't be a tuple.
+            For example: the following is not allowed:
+                for x in (1, 2.0, "string"):
+                    ....
+    """
+    iter_type = expr.infer(node.iter, context)
+    if not isinstance(iter_type, (TList, TSet, TIterator, TBytesString, TDictionary, TString)):
+        raise TypeError("The iterable should only be a set, list or iterator. Found {}.".format(iter_type))
+
+    # Infer the target in the loop, inside the global context
+    # Cases:
+    # - Var name. Ex: for i in range(5)..
+    # - Tuple. Ex: for (a,b) in [(1,"st"), (3,"st2")]..
+    # - List. Ex: for [a,b] in [(1, "st"), (3, "st2")]..
+    value_type = iter_type
+    if isinstance(iter_type, (TList, TSet, TIterator)):
+        value_type = value_type.type
+    elif isinstance(iter_type, TDictionary):
+        value_type = value_type.key_type
+    _infer_assignment_target(node.target, context, value_type)
+
+    return _infer_control_flow(node, context)
+
+def _infer_with(node, context):
+    """Infer the types for a with block"""
+    for item in node.items:
+        if item.optional_vars:
+            item_type = expr.infer(item.context_expr, context)
+            _infer_assignment_target(item.optional_vars, context, item_type)
+
+    return _infer_body(node.body, context)
+
+def _infer_try(node, context):
+    """Infer the types for a try/except/else block"""
+    try_type = UnionTypes()
+
+    try_type.union(_infer_body(node.body, context))
+    try_type.union(_infer_body(node.orelse, context))
+    try_type.union(_infer_body(node.finalbody, context))
+    # TODO: Infer exception handlers as classes
+
+    for handler in node.handlers:
+        try_type.union(_infer_body(handler.body, context))
+
+    if len(try_type.types) == 1:
+        return list(try_type.types)[0]
+    return try_type
+
+def infer(node, context):
+    if isinstance(node, ast.Assign):
+        return _infer_assign(node, context)
+    elif isinstance(node, ast.AugAssign):
+        return _infer_augmented_assign(node, context)
+    elif isinstance(node, ast.Return):
+        return expr.infer(node.value, context)
+    elif isinstance(node, ast.Delete):
+        return _infer_delete(node, context)
+    elif isinstance(node, (ast.If, ast.While)):
+        return _infer_control_flow(node, context)
+    elif isinstance(node, ast.For):
+        return _infer_for(node, context)
+    elif sys.version_info[0] >= 3 and sys.version_info[1] >= 5 and isinstance(node, ast.AsyncFor):
+        # AsyncFor is instroduced in Python 3.5
+        return _infer_for(node, context)
+    elif isinstance(node, ast.With):
+        return _infer_with(node, context)
+    elif sys.version_info[0] >= 3 and sys.version_info[1] >= 5 and isinstance(node, ast.AsyncWith):
+        # AsyncWith is instroduced in Python 3.5
+        return _infer_with(node, context)
+    elif isinstance(node, ast.Try):
+        return _infer_try(node, context)
+    return TNone()
```

Comment:
are we sure that want to return TNone() here? what about other statements not covered by the inference (assert, pass, etc.)?